---
title: "WGCNA Brain Ovaries"
output: html_notebook
editor_options: 
  chunk_output_type: inline
---
```{r packages, message=FALSE, warning=FALSE}
library(data.table)
library(tidyverse)
library(WGCNA)
library(viridis)
library(KEGGprofile)
library(DESeq2)
library(RColorBrewer)
```

#Load in Data from WGCNA Brain Ovaries file
```{r}
load("../data/dds_gene_counts.Rdata")

```

```{r DEG and PSGE overlaps}
PSGE_active<-read.csv("../data/PGSE-active.csv")%>%select("ID")
PSGE_sterile<-read.csv("../data/PGSE-sterile.csv")%>%select("ID")
DEG_reproductive<-read.csv("../data/DEG_reproductivestatus_galbraith2016.csv")%>%select("ID")

```

```{r}
active_brain<-read.table("../data/activeReadCounts_gene_brain.txt",header=T)%>%as.data.frame()
sterile_brain<-read.table("../data/sterileReadCounts_gene_brain.txt",header=T)%>%as.data.frame() 


```

```{r WGCNA ovary}
options(stringAsFactors = FALSE)
enableWGCNAThreads()

# We work with two sets:
nSets = 1;

# For easier labeling of plots, create a vector holding descriptive names of the two sets.
setLabels = c( "ovaries")
shortLabels = c("FBO")
# Form multi-set expression data: columns starting from 9 contain actual expression data.
ovaryExpr = vector(mode = "list", length = nSets)

###use ntd_brain, vst_brain,  rlog_brain, or common_brains
##use ntd_FBO, vst_FBO, rlog_FBO, or common_ovaries



ovaryExpr[[1]] = list(data=t(vst_all_FBO))

for(i in 1:nSets){
    ovaryExpr[[i]]$data<- apply(ovaryExpr[[i]]$data,2,as.numeric)
}

names(ovaryExpr[[1]]$data) <- rownames(vst_all_FBO)
rownames(ovaryExpr[[1]]$data) = colnames(vst_all_FBO)



exprSize = checkSets(ovaryExpr)


# Form a multi-set structure that will hold the clinical traits.
ovary_Traits = vector(mode="list", length = nSets);
for (set in 1:nSets)
{
setSamples = rownames(ovaryExpr[[set]]$data);
traitRows = match(setSamples, rownames(all_info));
ovary_Traits[[set]] = list(data = all_info[traitRows, ]);
rownames(ovary_Traits[[set]]$data) =rownames(all_info[traitRows, ]);
}

collectGarbage();
set.seed(10)
# Choose a set of soft-thresholding powers
powers = c(seq(4,10,by=1), seq(12,20, by=2));
# Initialize a list to hold the results of scale-free analysis
powerTables = vector(mode = "list", length = nSets);
# Call the network topology analysis function for each set in turn
for (set in 1:nSets)
powerTables[[set]] = list(data = pickSoftThreshold(ovaryExpr[[set]]$data, powerVector=powers,
verbose = 2)[[2]])

collectGarbage();
# Plot the results:
colors = c("black", "red", "blue", "green")
# Will plot these columns of the returned scale free analysis tables
plotCols = c(2,5,6,7)
colNames = c("Scale Free Topology Model Fit", "Mean connectivity", "Median connectivity",
"Max connectivity");
# Get the minima and maxima of the plotted points
ylim = matrix(NA, nrow = 2, ncol = 4);
for (set in 1:nSets)
{
  for (col in 1:length(plotCols))
  {
    ylim[1, col] = min(ylim[1, col], powerTables[[set]]$data[, plotCols[col]], na.rm = TRUE);
    ylim[2, col] = max(ylim[2, col], powerTables[[set]]$data[, plotCols[col]], na.rm = TRUE);
  }
}
# Plot the quantities in the chosen columns vs. the soft thresholding power
sizeGrWindow(8, 6)
pdf(file = "../figures/scaleFreeAnalysis.pdf", wi = 8, he = 6)
par(mfcol = c(2,2));
par(mar = c(4.2, 4.2 , 2.2, 0.5))
cex1 = 0.7;
for (col in 1:length(plotCols)) for (set in 1:nSets)
{
  if (set==1)
  {
    plot(powerTables[[set]]$data[,1], -sign(powerTables[[set]]$data[,3])*powerTables[[set]]$data[,2],
         xlab="Soft Threshold (power)",ylab=colNames[col],type="n", ylim = ylim[, col],
         main = colNames[col]);
    addGrid();
  }
  if (col==1)
  {
    text(powerTables[[set]]$data[,1], -sign(powerTables[[set]]$data[,3])*powerTables[[set]]$data[,2],
         labels=powers,cex=cex1,col=colors[set]);
  } else
    text(powerTables[[set]]$data[,1], powerTables[[set]]$data[,plotCols[col]],
         labels=powers,cex=cex1,col=colors[set]);
  if (col==1)
  {
    legend("bottomright", legend = setLabels, col = colors, pch = 20) ;
  } else
    legend("topright", legend = setLabels, col = colors, pch = 20) ;
}
dev.off()


net_ovaries = blockwiseConsensusModules(
ovaryExpr, power = 10, minModuleSize = 30, deepSplit = 2,
pamRespectsDendro = FALSE,
 corType="bicor",
mergeCutHeight = 0.25, numericLabels = TRUE,
minKMEtoStay = 0,nThreads = 3,
saveIndividualTOMs = T, saveConsensusTOMs = F, verbose = 5,
networkType = "signed", maxBlockSize = 14086,
TOMType = "signed")#can set block size to 15K for single analysis

ovaryMEs = net_ovaries$multiMEs
ovaryMLabels =net_ovaries$colors;
# Convert the numeric labels to color labels
ovaryMColors = labels2colors(ovaryMLabels)
ovaryTree = net_ovaries$dendrograms[[1]];


#consensus modules and colors
pdf(file = "../figures/Ovary Dendrogram w colors.pdf",width = 8, height = 6)
plotDendroAndColors(ovaryTree, ovaryMColors[net_ovaries$goodGenes],"Module colors",
dendroLabels = FALSE, hang = 0.03,
addGuide = FALSE, guideHang = 0.05,
main = "Ovary gene dendrogram and module colors")
dev.off()


 save(ovaryMEs, ovaryMLabels,ovaryMColors,ovaryTree,net_ovaries,ovary_Traits,ovaryExpr,setLabels,shortLabels, file = "../data/OvaryModules.RData")
# load("../data/OvaryModules.RData")




```


```{r WGCNA brain}
options(stringAsFactors = FALSE)
enableWGCNAThreads()

# We work with two sets:
nSets = 1;

# For easier labeling of plots, create a vector holding descriptive names of the two sets.
setLabels = c( "Brain")
shortLabels = c("B")
# Form multi-set expression data: columns starting from 9 contain actual expression data.
brainExpr = vector(mode = "list", length = nSets)

###use ntd_brain, vst_brain,  rlog_brain, or common_brains
##use ntd_FBO, vst_FBO, rlog_FBO, or common_ovaries



brainExpr[[1]] = list(data=t(vst_all_brain))

for(i in 1:nSets){
    brainExpr[[i]]$data<- apply(brainExpr[[i]]$data,2,as.numeric)
}

names(brainExpr[[1]]$data) <- rownames(vst_all_brain)
rownames(brainExpr[[1]]$data) <- colnames(vst_all_brain)



exprSize = checkSets(brainExpr)


# Form a multi-set structure that will hold the clinical traits.
brain_Traits = vector(mode="list", length = nSets);
for (set in 1:nSets)
{
setSamples = rownames(brainExpr[[set]]$data);
traitRows = match(setSamples, rownames(all_info));
brain_Traits[[set]] = list(data = all_info[traitRows, ]);
rownames(brain_Traits[[set]]$data) =rownames(all_info[traitRows, ]);
}

collectGarbage();

net_brain = blockwiseConsensusModules(
brainExpr, power = 10, minModuleSize = 30, deepSplit = 2,
pamRespectsDendro = FALSE,
 corType="bicor",
mergeCutHeight = 0.25, numericLabels = TRUE,
minKMEtoStay = 0,nThreads = 3,
saveIndividualTOMs = T, saveConsensusTOMs = T, verbose = 5,
networkType = "signed", maxBlockSize = 14086,
TOMType = "signed")#can set block size to 15K for single analysis

brainMEs = net_brain$multiMEs
brainMLabels =net_brain$colors;
# Convert the numeric labels to color labels
brainMColors = labels2colors(brainMLabels)
brainTree = net_brain$dendrograms[[1]];


#consensus modules and colors
pdf(file = "../figures/Brain Dendrogram w colors.pdf",width = 8, height = 6)

plotDendroAndColors(brainTree, brainMColors[net_brain$goodGenes],"Module colors",
dendroLabels = FALSE, hang = 0.03,
addGuide = FALSE, guideHang = 0.05,
main = "Brain gene dendrogram and module colors")
dev.off()


 save(brainMEs, brainMLabels,brainMColors,brainTree,net_brain,brain_Traits,brainExpr,setLabels,shortLabels, file = "../data/BrainModules.RData")
load("../data/BrainModules.RData")
 



```

```{r glm ovary }
load("../data/OvaryModules.RData")

##setup variables
nSets=1
moduleTraitCor_ovary = list();
moduleTraitPvalue_ovary = list();


# Calculate the correlations

for (set in 1:nSets){
    
res <- data.frame()
pvalues<-data.frame()
ME_names_ovary<-colnames(ovaryMEs[[1]]$data)

for (i in ME_names_ovary){

    trait_i<-ovary_Traits[[set]]$data
  test<-glm(ovaryMEs[[set]]$data[,i]~trait_i$MatxPat+
                trait_i$Block+
                trait_i$Rep_status)
  #test%>%summary()%>%print()
  ##get coefficients
  coeff2<-coef(test)[2]
  coeff3<-coef(test)[3]
  coeff4<-coef(test)[4]
  coeff<-data.frame(MatxPat=coeff2,
                    Block=coeff3,
                    Rep_status=coeff4,
                    row.names = i)
  ##get pvalues
  pval2<-coef(summary(test))[2,4]
  pval3<-coef(summary(test))[3,4]
  pval4<-coef(summary(test))[4,4]
  pval<-data.frame(MatxPat=pval2,
                   Block=pval3,
                   Rep_status=pval4,
                   row.names = i)
  pvalues<-rbind(pvalues,pval)
  res<-rbind(res,coeff)
  }
moduleTraitCor_ovary[[set]]<-res
moduleTraitPvalue_ovary[[set]]<-pvalues
}
```


```{r plotting colors}
###Build consensus figure
 heat.col<-viridis(100)
pal<- brewer.pal(11,"RdBu") #or "BrBG"

cols <- c(colorRampPalette(c(pal[11], pal[6]))(51), 
    colorRampPalette(c(pal[6], pal[1]))(51)[-1])
```





```{r ovary modules figure signed}


# Initialize matrices to hold the consensus correlation and p-value
OvaryCor = matrix(NA, nrow(moduleTraitCor_ovary[[1]]), ncol(moduleTraitCor_ovary[[1]]));
OvaryPvalue = matrix(NA, nrow(moduleTraitCor_ovary[[1]]), ncol(moduleTraitCor_ovary[[1]]));
# Find consensus negative correlations
negative = moduleTraitCor_ovary[[1]] < 0 
OvaryCor[negative] = pmax(moduleTraitCor_ovary[[1]][negative]);
OvaryPvalue[negative] = pmax(moduleTraitPvalue_ovary[[1]][negative]);
# Find consensus positive correlations
positive = moduleTraitCor_ovary[[1]] > 0 
OvaryCor[positive] = pmin(moduleTraitCor_ovary[[1]][positive]);
OvaryPvalue[positive] = pmax(moduleTraitPvalue_ovary[[1]][positive]);

##plotting consensus figure

OvaryPvalue[is.na(OvaryPvalue)]<-1
OvaryPvalue<-as.data.frame(OvaryPvalue)
##only display those that are significant
display<-pmap_dfr(OvaryPvalue,~data.frame(min_MatxPat=..1>0.05 , min_Block=..2>0.05 , min_Rep=..3>0.05))

OvaryCor[display==T]<-NA

pvalues<-OvaryPvalue
pvalues<-cbind(pvalues,pmap_dfr(pvalues,~data.frame(min=min(..1,..2,..3))))

# keep only those modules with at least 1 pvalue <0.05
# For brain or ovary alone, it looks better to run the following code to filter figure
# However to build consensus figure, don't filter yet to keep number of rows the same
rownames(OvaryCor)<-rownames(moduleTraitCor_ovary[[1]])
colnames(OvaryCor)<-colnames(moduleTraitCor_ovary[[1]])

OvaryPvalue<-OvaryPvalue[pvalues$min<0.05,]
OvaryCor<-OvaryCor[pvalues$min<0.05,]


textMatrix = paste(signif(OvaryCor, 2)%>%as.matrix(), "\n(P = ",
signif(OvaryPvalue, 2)%>%as.matrix(), ")", sep = "")
dim(textMatrix) = dim(OvaryCor)


pdf(file = "../figures/OvaryTraitRelationships-signed.pdf", wi = 6, he = 6);
par(mar = c(6, 8.8, 3, 2.2));
labeledHeatmap(Matrix = OvaryCor,
xLabels = colnames(OvaryCor),
yLabels = rownames(OvaryCor),
ySymbols = rownames(OvaryCor),
colorLabels = FALSE,
colors = cols,
textMatrix = textMatrix,
setStdMargins = FALSE,
cex.text = 0.5,
zlim = c(-1,1),
main = paste(" module-trait relationships across\n",
paste(setLabels, collapse = " and ")))
dev.off()
```

```{r glm brain }
load("../data/BrainModules.RData")

##setup variables
nSets=1
moduleTraitCor_brain = list();
moduleTraitPvalue_brain = list();


# Calculate the correlations

for (set in 1:nSets){
    
res <- data.frame()
pvalues<-data.frame()
ME_names_brain<-colnames(brainMEs[[1]]$data)

for (i in ME_names_brain){
  trait_i<-brain_Traits[[set]]$data
  test<-glm(brainMEs[[set]]$data[,i]~trait_i$MatxPat+
                trait_i$Block+
                trait_i$Rep_status)
  #test%>%summary()%>%print()
  ##get coefficients
  coeff2<-coef(test)[2]
  coeff3<-coef(test)[3]
  coeff4<-coef(test)[4]
  coeff<-data.frame(MatxPat=coeff2, 
                    Block=coeff3,
                    Rep_status=coeff4,
                    row.names = i)
  ##get pvalues
  pval2<-coef(summary(test))[2,4]
  pval3<-coef(summary(test))[3,4]
  pval4<-coef(summary(test))[4,4]
  pval<-data.frame(MatxPat=pval2,
                   Block=pval3,
                   Rep_status=pval4,
                   row.names = i)
  pvalues<-rbind(pvalues,pval)
  res<-rbind(res,coeff)
  }
moduleTraitCor_brain[[set]]<-res
moduleTraitPvalue_brain[[set]]<-pvalues
}
```




```{r brain modules figure signed}


# Initialize matrices to hold the consensus correlation and p-value
BrainCor = matrix(NA, nrow(moduleTraitCor_brain[[1]]), ncol(moduleTraitCor_brain[[1]]));
BrainPvalue = matrix(NA, nrow(moduleTraitCor_brain[[1]]), ncol(moduleTraitCor_brain[[1]]));
# Find consensus negative correlations
negative = moduleTraitCor_brain[[1]] < 0 
BrainCor[negative] = pmax(moduleTraitCor_brain[[1]][negative]);
BrainPvalue[negative] = pmax(moduleTraitPvalue_brain[[1]][negative]);
# Find consensus positive correlations
positive = moduleTraitCor_brain[[1]] > 0 
BrainCor[positive] = pmin(moduleTraitCor_brain[[1]][positive]);
BrainPvalue[positive] = pmax(moduleTraitPvalue_brain[[1]][positive]);

##plotting consensus figure

BrainPvalue[is.na(BrainPvalue)]<-1
BrainPvalue<-as.data.frame(BrainPvalue)
##only display those that are significant
display<-pmap_dfr(BrainPvalue,~data.frame(min_MatxPat=..1>0.05 , min_Block=..2>0.05 , min_Rep=..3>0.05))

BrainCor[display==T]<-NA

pvalues<-BrainPvalue
pvalues<-cbind(pvalues,pmap_dfr(pvalues,~data.frame(min=min(..1,..2,..3))))

# keep only those modules with at least 1 pvalue <0.05
# For brain or ovary alone, it looks better to run the following code to filter figure
# However to build consensus figure, don't filter yet to keep number of rows the same
rownames(BrainCor)<-rownames(moduleTraitCor_brain[[1]])
colnames(BrainCor)<-colnames(moduleTraitCor_brain[[1]])

BrainPvalue<-BrainPvalue[pvalues$min<0.05,]
BrainCor<-BrainCor[pvalues$min<0.05,]


textMatrix_brain = paste(signif(BrainCor, 2)%>%as.matrix(), "\n(P = ",
signif(BrainPvalue, 2)%>%as.matrix(), ")", sep = "")
dim(textMatrix_brain) = dim(BrainCor)


pdf(file = "../figures/BrainTraitRelationships-signed.pdf", wi = 6, he = 6);
par(mar = c(6, 8.8, 3, 2.2));
labeledHeatmap(Matrix = BrainCor,
xLabels = colnames(BrainCor),
yLabels = rownames(BrainCor),
ySymbols = rownames(BrainCor),
colorLabels = FALSE,
colors = cols,
textMatrix = textMatrix_brain,
setStdMargins = FALSE,
cex.text = 0.5,
zlim = c(-1,1),
main = paste(" module-trait relationships across\n",
paste(setLabels, collapse = " and ")))
dev.off()
```




```{r getting ovary module gene lists}
#consMEs,consMColors,consMLabels,ME_names
conversion<-read.csv("../data/mRNA_Entrez_Genbank_HAv3.1beebase_conversion.csv",header = T)%>%select(Beebase=OtherID, Entrez=GeneID)
MEdata_ovary<-ovaryExpr[[1]]$data
gene_name<-colnames(MEdata_ovary)%>%as.data.frame
colnames(gene_name)<-"Beebase"
gene_entrez<-left_join(gene_name,conversion)%>%distinct
all(gene_entrez$Beebase==colnames(MEdata_ovary))

##this line of code subs in the entrez, which is used for the KEGGs
#However, beebase useful for the overlaps with DMLs
#colnames(MEdata)<-gene_entrez$Entrez





module_ovary_significant_genes<-data.frame(background=colnames(MEdata_ovary))
for (i in ME_names_ovary){
  whichModule= sub(i,pattern = "ME",replacement = "")
  name<-paste(i,"_kME",sep = "")
  
genes<-row.names(t(MEdata_ovary))[ovaryMLabels==whichModule]
length(genes)<-length(row.names(t(MEdata_ovary)))

name=data.frame("gene"=genes)
module_ovary_significant_genes<-cbind(module_ovary_significant_genes,genes)
}
colnames(module_ovary_significant_genes)<-c("background",ME_names_ovary)

write.csv(module_ovary_significant_genes, "../output/WGCNA_ovaryspecific_Sig_Modules_genes_entrez.csv")

##Getting Hub Genes
power=10

Hubgenes_FBO<-chooseTopHubInEachModule(datExpr = ovaryExpr[[1]]$data, #gene expression data, rows =samples; columns =genes  
   colorh=ovaryMColors, #module assignments for all rows  
   omitColors = "grey", 
   power = power, 
   type = "signed")

```

```{r getting brain module gene lists}
#consMEs,consMColors,consMLabels,ME_names
conversion<-read.csv("../data/mRNA_Entrez_Genbank_HAv3.1beebase_conversion.csv",header = T)%>%select(Beebase=OtherID, Entrez=GeneID)
MEdata_brain<-brainExpr[[1]]$data
gene_name<-colnames(MEdata_brain)%>%as.data.frame
colnames(gene_name)<-"Beebase"
gene_entrez<-left_join(gene_name,conversion)%>%distinct
all(gene_entrez$Beebase==colnames(MEdata_brain))

##this line of code subs in the entrez, which is used for the KEGGs
#However, beebase useful for the overlaps with DMLs
#colnames(MEdata)<-gene_entrez$Entrez





module_brain_significant_genes<-data.frame(background=colnames(MEdata_brain))
for (i in ME_names_brain){
  whichModule= sub(i,pattern = "ME",replacement = "")
  name<-paste(i,"_kME",sep = "")
  
genes<-row.names(t(MEdata_brain))[brainMLabels==whichModule]
length(genes)<-length(row.names(t(MEdata_brain)))

name=data.frame("gene"=genes)
module_brain_significant_genes<-cbind(module_brain_significant_genes,genes)
}
colnames(module_brain_significant_genes)<-c("background",ME_names_brain)

write.csv(module_brain_significant_genes, "../output/WGCNA_brainspecific_Sig_Modules_genes_entrez.csv")

##Getting Hub Genes
power=10

Hubgenes_FBO<-chooseTopHubInEachModule(datExpr = ovaryExpr[[1]]$data, #gene expression data, rows =samples; columns =genes  
   colorh=ovaryMColors, #module assignments for all rows  
   omitColors = "grey", 
   power = power, 
   type = "signed")

```


```{r DEG sanity check gene lists}
all_mod_genes<-ovaryExpr[[1]]$data%>%colnames()
overlaps_DEG_background<-intersect(DEG_reproductive$ID,all_mod_genes)
DEG_reproductive<-DEG_reproductive$ID
```


```{r DEGs for brain}
source("../code/get_significant_genes_function.R")

##do the differential expression
DEG_brain<-DESeq(dds_all_brain)

##extract results
results_rep_status<-results(DEG_brain, contrast =c("Rep_status","Sterile","Reproductive"),alpha = 0.05)

results_matxpat<-results(DEG_brain, contrast =c("MatxPat","ExA","AxE"),alpha = 0.05)

results_block<-results(DEG_brain, contrast =c("Block","A","B"),alpha = 0.05)

#display results summaries
results_rep_status%>%summary()
results_matxpat%>%summary()
results_block%>%summary()

#extract the gene names
results_rep_status%>%get_sig_genes()->rep_stat_idx
results_matxpat%>%get_sig_genes()->matxpat_idx
results_block%>%get_sig_genes()->block_idx
```


```{r brain overlaps DMLs}
#read in DMLS and get gene lists immediately
active_parentDML<-read.csv(file = "../DMLgene_lists_05-01-19/rep_parent_DMLgenes.csv")%>%select(V9)
active_strainDML<-read.csv(file = "../DMLgene_lists_05-01-19/rep_strain_DMLgenes.csv")%>%select(V9)
sterile_parentDML<-read.csv(file = "../DMLgene_lists_05-01-19/sterile_parent_DMLgenes.csv")%>%select(V9)
sterile_strainDML<-read.csv(file = "../DMLgene_lists_05-01-19/sterile_strain_DMLgenes.csv")%>%select(V9)


#read in beebase ids
mod_sig_beebase<-read.csv("../output/WGCNA_ovaryspecific_Sig_Modules_genes_entrez.csv")%>%select(-X)

#name of modules
module_name<-mod_sig_beebase%>%colnames()

#get number of gene overlaps
m_intersect <- function(var1,var2){intersect(var1,var2)%>%unique()%>%length()}

#get number of unique genes
unique_length<-function(x){length(unique(x))}

#get size of each module
module_size<-apply(X = mod_sig_beebase,2, FUN = unique_length)

##pvalues set in individ figures
## is it in the figure? included extra NA for background

included_ovaries_fig<-c(NA,pvalues$min<0.05)

#get overlaps for active and sterile
active_parent_overlap<-sapply(mod_sig_beebase,m_intersect,var2=active_parentDML$V9)
active_strain_overlap<-sapply(mod_sig_beebase,m_intersect,var2=active_strainDML$V9)

sterile_parent_overlap<-sapply(mod_sig_beebase,m_intersect,var2=sterile_parentDML$V9)
sterile_strain_overlap<-sapply(mod_sig_beebase,m_intersect,var2=sterile_strainDML$V9)

moduleDML_overlaps<-cbind(module_name,module_size,active_parent_overlap,active_strain_overlap,sterile_parent_overlap,sterile_strain_overlap)



#Hypergeometric tests for active genes

##length active parent
act_parent_length<-active_parentDML$V9%>%unique_length()

##length active strain
act_strain_length<-active_strainDML$V9%>%unique_length()

##sterile parent
ster_parent_length<-sterile_parentDML$V9%>%unique_length()

##sterile strain
ster_strain_length<-sterile_strainDML$V9%>%unique_length()

background<-12836

#active parent overlaps
active_parentDML_pvalue<-c()
for(mod in 1:length(module_name)){
active_parentDML_pvalue[mod]<-1-phyper(active_parent_overlap[mod], act_parent_length, background-act_parent_length, module_size[mod], lower.tail = TRUE, log.p = FALSE)
}


##active strain overlaps
active_strainDML_pvalue<-c()
for(mod in 1:length(module_name)){
active_strainDML_pvalue[mod]<-1-phyper(active_strain_overlap[mod], act_strain_length, background-act_strain_length, module_size[mod], lower.tail = TRUE, log.p = FALSE)
}

##sterile strain overlaps
ster_strainDML_pvalue<-c()
for(mod in 1:length(module_name)){
ster_strainDML_pvalue[mod]<-1-phyper(sterile_strain_overlap[mod], ster_strain_length, background-ster_strain_length, module_size[mod], lower.tail = TRUE, log.p = FALSE)
}


#sterile parent overlaps
ster_parentDML_pvalue<-c()
for(mod in 1:length(module_name)){
ster_parentDML_pvalue[mod]<-1-phyper(sterile_parent_overlap[mod], ster_parent_length, background-ster_parent_length, module_size[mod], lower.tail = TRUE, log.p = FALSE)
}


##get overlaps for DEGs
active_PSGE_module_overlap<-sapply(mod_sig_beebase,m_intersect,var2=PSGE_active$ID)

sterile_PSGE_module_overlap<-sapply(mod_sig_beebase,m_intersect,var2=PSGE_sterile$ID)

DEG_reproductive_module_overlap<-sapply(mod_sig_beebase,m_intersect,var2=DEG_reproductive)

#length of genes
sterile_PSGE_length<-PSGE_sterile$ID%>%unique_length()##length active psge
active_PSGE_length<-PSGE_active$ID%>%unique_length()#length sterile psge
DEG_length<-DEG_reproductive%>%unique_length()#length deg

#Sterlie psge oerlaps
sterilePSGE_repstat_pvalue<-c()
for(mod in 1:length(module_name)){
sterilePSGE_repstat_pvalue[mod]<-1-phyper(sterile_PSGE_module_overlap[mod], sterile_PSGE_length, background-sterile_PSGE_length, module_size[mod], lower.tail = TRUE, log.p = FALSE)
}

#active psge overlaps
activePSGE_repstat_pvalue<-c()
for(mod in 1:length(module_name)){
activePSGE_repstat_pvalue[mod]<-1-phyper(active_PSGE_module_overlap[mod], active_PSGE_length, background-active_PSGE_length, module_size[mod], lower.tail = TRUE, log.p = FALSE)
}

#DEG_opverlaps
DEG_repstat_pvalue<-c()
for(mod in 1:length(module_name)){
DEG_repstat_pvalue[mod]<-1-phyper(DEG_reproductive_module_overlap[mod], DEG_length, background-DEG_length, module_size[mod], lower.tail = TRUE, log.p = FALSE)
}

moduleDEG_PSGE_genes_results<-cbind(module_name,module_size,active_PSGE_module_overlap,sterile_PSGE_module_overlap,DEG_reproductive_module_overlap,activePSGE_repstat_pvalue,sterilePSGE_repstat_pvalue,DEG_repstat_pvalue)%>%as.data.frame()

#creating data sets
moduleDMLgenes_results<-cbind(moduleDML_overlaps,active_parentDML_pvalue,active_strainDML_pvalue, ster_parentDML_pvalue, ster_strainDML_pvalue)%>%as.data.frame()
all(moduleDEG_PSGE_genes_results$module_name==moduleDMLgenes_results$module_name)

#combing DEG,DML,PSGE
module_DEG_DML_PSGE_results<-left_join(moduleDMLgenes_results,moduleDEG_PSGE_genes_results)

##adding in reproductive status pvalue from GLM
ovary_pvalues<-moduleTraitPvalue_ovary[[1]]$Rep_status%>%as.data.frame()
ovary_pvalues$module_name<-rownames(moduleTraitPvalue_ovary[[1]])
ovary_pvalues$ovary_correlation<-moduleTraitCor_ovary[[1]]$Rep_status
colnames(ovary_pvalues)<-c("reproductive_pvalue", "module_name", "ovary_correlation")

##combining_everything
module_DEG_DML_PSGE_results<-left_join(module_DEG_DML_PSGE_results, ovary_pvalues)


write.csv(moduleDMLgenes_results,"../output/ovary_moduleDMLoverlap_results.csv")
write.csv(moduleDEG_PSGE_genes_results,"../output/ovary_moduleDEG_PSGE_overlap_results.csv")
write.csv(module_DEG_DML_PSGE_results,"../output/ovary_module_DEG_DML_PSGE_overlap_results.csv") 

##looking at specific modules
# active_DEG_module42_overlap<-intersect(mod_sig_beebase$ME42,PSGE_active )
# sterile_DEG_module42_overlap<-intersect(mod_sig_beebase$ME42,PSGE_sterile )

```

```{r keggs for module genes}
source("../code/find_kegg.R")

kegg_pathways_WGCNA<-find_keggs(module_significant_genes)
colnames(kegg_pathways_WGCNA)<-colnames(module_significant_genes)


write.csv(kegg_pathways_WGCNA, file = "../output/kegg_pathways_WGCNA_BvFBO.csv")
#read.csv("../output/kegg_pathways_WGCNA_BvFBO.csv")
kegg_pathways_WGCNA$ME42
```


```{r exporting consensus to Cytoscape}
# Recalculate topological overlap if needed
load("../data/ConsensusModules_BvsFBO.RData")
Cons_name<-load("../data/consensusTOM-block.1.RData")

consTOM = consTomDS%>%as.matrix()
# Read in the annotation file
annot <-read.csv("../data/mRNA_Entrez_Genbank_HAv3.1beebase_conversion.csv",header = T)
# Select modules
modules =  "21";##select which modules you want to see
goodgenes<-net$goodGenes
# Select module probes
probes = colnames(multiExpr[[1]]$data)[goodgenes]
inModule = is.finite(match(consMLabels, modules))[goodgenes];
modProbes = probes[inModule];
modGenes = annot$GeneID[match(modProbes, annot$OtherID)][]
# Select the corresponding Topological Overlap
modTOM = consTOM[inModule, inModule];


dimnames(modTOM) = list(modProbes, modProbes)
# Export the network into edge and node list files Cytoscape can read
cyt = exportNetworkToCytoscape(modTOM,
edgeFile = paste("../network_vis_for_cytoscape/CytoscapeInput-edges-", paste(modules, collapse="-"), ".txt", sep=""),
nodeFile = paste("../network_vis_for_cytoscape/CytoscapeInput-nodes-", paste(modules, collapse="-"), ".txt", sep=""),
weighted = TRUE,
threshold = 0.02,
nodeNames = modProbes,
altNodeNames = modGenes,
nodeAttr = net$colors[goodgenes][inModule]);
```


